### ISSUE TITLE
[WIP] [spec] Automating Case Storage assignment to judges

### ISSUE THREAD

**Issue by amprokop** [department-of-veterans-affairs/caseflow#4855]
# Context

See https://github.com/department-of-veterans-affairs/appeals-design-research/issues/645

# Overview

We'll develop backend logic that removes any need for a Case Storage staff member to push cases to judges. By this, we'll eliminate any lag time and ensure judges can get cases at will.

# Implementation

When an appeal is in case storage, the value for `BRIEFF.BFCURLOC` is 81. To assign N appeals to a judge by docket order, we'd:

- Get the current docket date using `VACOLS::CaseDocket.docket_date_of_nth_appeal_in_case_storage` 
- Find the oldest appeals in location 81 
   - where BFD19 <= current docket date
   - that are activated
   - that don't have a hearing date
   - that have no pending mail or open diaries
- Update those N appeals to the judge's location
- Create DECASS rows for each appeal where DEFOLDER is equal to BRIEFF.BFKEY

After that, a subsequent call to `/queue/:user_id` will fetch the updated list of appeals assigned to the judge.

We'll weigh two different approaches for triggering the assignment of cases from case storage to judges.

### Option 1: Self-service button

When a judge loads Caseflow Queue and views the appeals assigned to them, we could present the judge a "Fetch more from Case Storage" button. Upon clicking a button, we'd assign a batch of case storage cases to the judge, and they'd appear in the Queue list view immediately after assignment is complete.

#### Goal: Judges should understand the methodology behind automatic case allocation
  - The button would require minimal training and explanation
#### Goal: Cases should not be allocated to attorneys and judges on PTO/leave
  - Judges are likely in the best position to know who is on PTO/leave, and they won't request cases for themselves while they're on leave
#### Goal: Cases are only allocated to active staff 
  - Cases will not be assigned to judges unless they click a button, and judges will know not to assign cases to inactive attorneys
#### Goal: Cases should not sit on judges plates to assign or to sign for long times
  - A "Fetch more from Case Storage" button could be disabled until after a judge has assigned all existing cases 

### Option 2: Auto-assign cases to judges from case storage

We wouldn't need any UI for this option. We could approach a full automation option in several ways. Two that come to mind:

1. A recurring job could run on some interval and ensure each judge has at least N cases on their plate to assign to attorneys. 

2. Each time a judge assigns an appeal to an attorney, if the judge has less than N cases on their plate, we add a batch of new cases from case storage

However, a fully automated option will create some challenges and risks.

#### Goal: Judges should understand the methodology behind automatic case allocation
  - We would need to think about how to ensure that it doesn't feel like cases are assigned at random times
#### Goal: Cases should not be allocated to attorneys and judges on PTO/leave
 - Accounting for PTO/leave may prove to be difficult and error-prone— delays may result
#### Goal: Cases are only allocated to active staff 
  - We would need to ensure that we don’t accidentally assign cases to judges who have retired, and if cases do get assigned to those judges, we should audit it.
#### Goal: Cases should not sit on judges plates to assign or to sign for long times
  - If their plates are automatically refilled, judges might hold on to difficult cases longer than they would otherwise

# Test Plan

Using FACOLS, we'll be able to thoroughly test the Case Storage assignment to judge logic locally. Much of the underlying components are already written.

# Rollout Plan

We can develop this in parallel with the judge queue, but we shouldn't roll it out until we roll out the judge queue. We could test this process change with one judge at first by putting it behind a feature toggle.

**Comment by laurjpeterson** [comment]
Thanks for writing this up!

## Re: Implementation section. 
Today case storage assigns AOD cases and general population cases (non-hearing or retired judge cases where the Veteran opted not to have a new hearing) to senior counsel, who then assign those cases as evenly as possible among judges. They try not to give too many AOD cases to judges at once. I think automatic case assignment could deprecate this manual step. 

Automatic case assignment should prioritize assigning judges their legacy hearing cases, but also take into account AOD and general population cases. Depending on how big certain judges' backlog of their own hearing cases are that are within the docket range, some judges might not get any AOD or gen pop cases in each auto-assignment, while other judges have too many AODs. 

Could we add the hard requirements to the implementation above, then potentially test with real data how AODs and gen pops are distributed across judges?

Requirements that should be taken into account:
- Hard requirements:
  - Cases must be assigned in docket order, oldest to newest form 9 dates.
  - AOD trumps docket order
  - Judges must be assigned their own legacy hearing cases 
- Not hard, but should prioritize:
- Nice to have requirements:
  - Where possible, attorneys receive cases they have already touched e.g. for post remand cases. Could we identify former attorney, match to their judge, and assign to that judge. We could only do this if it was not a hearing case, or if that same judge conducted the Veteran's hearing. 

## Re: triggering
After reading through, I initially like the option 1: self-service button, primarily for the PTO/out of office concerns. I especially like disabling the button until they have finished assigning out cases.

**Comment by amprokop** [comment]
Thanks for the additional detail! 

For Option 1, a few details about the self-service button to research/consider later:
 - Should it assign a constant # of cases? Say, 20 at a time. (suggest we start here and see what judges ask for)
- Otherwise, it could assign X cases for every Y attorneys in the judge's group
- Or we could let the judge choose a number of cases to request in the UI 

Re: prioritization, two questions come to mind:

- Will we create a logjam if we always prioritize hearing cases before all other cases?
- Will we create a logjam if we always prioritize AOD cases before all other cases?

Hypotheses that I'd like to test: no to 1, yes to 2.

**Comment by laurjpeterson** [comment]
For Option 1 - agree with starting between 20-30 cases at a time. Our second round of research confirmed that when judges have that amount to assign at a given time. Your second bullet, x cases for every y attorney, is appealing down the line, though. Especially if there are more attorneys on judge teams in the future.

**Comment by laurjpeterson** [comment]
most up to date version: https://github.com/department-of-veterans-affairs/caseflow/issues/4985
